planner/core/common_plans.go

```diff
@@ -451,9 +451,22 @@ func (e *Execute) getPhysicalPlan(ctx context.Context, sctx sessionctx.Context,
 	stmtCtx.UseCache = prepared.UseCache
 
 	var bindSQL string
+
+	schemaCheck := sctx.GetSessionVars().IsIsolation(ast.ReadCommitted) || preparedStmt.ForUpdateRead
 	if prepared.UseCache {
 		bindSQL = GetBindSQL4PlanCache(sctx, preparedStmt)
-		if cacheKey, err = NewPlanCacheKey(sctx.GetSessionVars(), preparedStmt.StmtText, preparedStmt.StmtDB, prepared.SchemaVersion); err != nil {
```

Comment:
I think we should add a field with the latest schema version to compute the cacheKey, instead of replacing the old one. Or sometimes the newer txn will use the old one's cache. For example:

```sql
create table t(id int);
insert into t values(1);
prepare s from 'select * from t for update';
begin optimistic;
              alter table t add column (v int); --- another session add a new column v
execute s; --- result: 1, and the plan will be cached
rollback;
begin optimistic;
execute s; --- Should be (1, NULL), but the second txn is using the cached plan, so only one column is returned in fact
```

Reply:
I have fixed it and add a test. Thanks.